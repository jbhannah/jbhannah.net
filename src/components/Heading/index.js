import styled from "@emotion/styled"
import PropTypes from "prop-types"
import React from "react"

const styles = {
  "& a": {
    background: "none",
  },
}

const headings = {
  h1: styled.h1(styles),
  h2: styled.h2(styles),
  h3: styled.h3(styles),
  h4: styled.h4(styles),
  h5: styled.h5(styles),
  h6: styled.h6(styles),
}

const Heading = ({ children, level, ...props }) => {
  const H = headings[level]
  return <H {...props}>{children}</H>
}

export default Heading

Heading.propTypes = {
  children: PropTypes.node,
  level: PropTypes.oneOf(Object.keys(headings)),
}
