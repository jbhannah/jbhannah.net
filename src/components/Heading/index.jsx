import PropTypes from "prop-types"
import React from "react"

const styles = {
  a: {
    background: "none",
    textShadow: "none",
  },
  code: {
    fontSize: "inherit",
  },
}

export const Heading = ({ level, ...props }) => {
  const H = level
  return <H css={styles} {...props} />
}

export const HeadingFactory = level => props => (
  <Heading level={level} {...props} />
)

export default Heading

Heading.propTypes = {
  level: PropTypes.oneOf(["h1", "h2", "h3", "h4", "h5", "h6"]),
}
