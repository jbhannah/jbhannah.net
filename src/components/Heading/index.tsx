import * as React from "react"

const headingCSS = {
  a: {
    background: "none",
    textShadow: "none",
  },
  code: {
    fontSize: "inherit",
  },
}

type HeadingLevel = "h1" | "h2" | "h3" | "h4" | "h5" | "h6"

type HeadingBaseProps = JSX.IntrinsicAttributes &
  React.ClassAttributes<HTMLHeadingElement> &
  React.HTMLAttributes<HTMLHeadingElement>

interface HeadingProps extends HeadingBaseProps {
  level: HeadingLevel
}

export const Heading: React.FunctionComponent<HeadingProps> = ({
  level: H,
  ...props
}) => {
  return <H css={headingCSS} {...props} />
}

export const HeadingFactory = (level: HeadingLevel) => (
  props: HeadingBaseProps
) => <Heading level={level} {...props} />

export const H1 = HeadingFactory("h1")
export const H2 = HeadingFactory("h2")
export const H3 = HeadingFactory("h3")
export const H4 = HeadingFactory("h4")
export const H5 = HeadingFactory("h5")
export const H6 = HeadingFactory("h6")
