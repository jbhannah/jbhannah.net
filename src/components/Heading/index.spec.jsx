import { shallow } from "enzyme"
import React from "react"
import Heading, { HeadingFactory } from "."

const TestHeading = (props) => shallow(<Heading {...props} />)

describe("Heading", () => {
  describe("with a valid heading level", () => {
    const tree = TestHeading({ level: "h1" })

    it("renders correctly", () => {
      expect(tree).toMatchSnapshot()
    })
  })

  describe("with an invalid heading level", () => {
    it("prints a console error", () => {
      const warn = jest.spyOn(console, "error")
      warn.mockImplementation()

      TestHeading({ level: "h7" })

      expect(warn).toBeCalled()
      warn.mockRestore()
    })
  })
})

describe("HeadingFactory", () => {
  const tree = shallow(HeadingFactory("h1")())

  it("returns a Heading", () => {
    expect(tree.is("h1")).toBeTruthy()
  })
})
