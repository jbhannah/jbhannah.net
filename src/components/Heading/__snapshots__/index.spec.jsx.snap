// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading with a valid heading level renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Heading
    level="h1"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "css": Object {
        "a": Object {
          "background": "none",
          "textShadow": "none",
        },
        "code": Object {
          "fontSize": "inherit",
        },
      },
    },
    "ref": null,
    "rendered": null,
    "type": "h1",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "css": Object {
          "a": Object {
            "background": "none",
            "textShadow": "none",
          },
          "code": Object {
            "fontSize": "inherit",
          },
        },
      },
      "ref": null,
      "rendered": null,
      "type": "h1",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
